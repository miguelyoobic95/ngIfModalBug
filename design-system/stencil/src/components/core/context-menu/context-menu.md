---
name: Context Menu
category: Core Components
---
Provides a context menu. Context menu entries comply with the `IContextMenuEntry` interface.

``` typescript
export interface IContextMenuEntry {
    itemTitle?: string;
    icon?: string;
    class?: string;
    active?: boolean;
    separator?: boolean;
    separatorAfter?: boolean;
    hidden?: boolean;
    isHidden?: Function;
    visible?: boolean;
    handler?: Function;
    items?: Array<IContextMenuEntry>;
    sendImmutable?: boolean;
}
```

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`items`|`IContextMenuEntry[]`|   |the context menu entries to display|
|`context`|any|   |an object to represent the menu's context |

```yoo-context-menu.html
<div style="padding:10px;">
    <yoo-context-menu>
        <yoo-button slot="trigger" class="squared" text="test"></yoo-button>
        <div>Lorem ipsum dolor sit amet consectetur adipisicing elit. Totam sequi vero laudantium cumque similique consectetur corporis blanditiis error neque ex libero, ipsa nostrum optio pariatur necessitatibus. Quas perferendis provident ipsam.</div>
    </yoo-context-menu>


     <yoo-context-menu id="contextmenu">
        <yoo-button class="accent squared" slot="trigger" text="test"></yoo-button>
    </yoo-context-menu>
</div>
```

```yoo-context-menu.js
    var cmp = document.querySelector('#contextmenu');
    cmp.items = [{
        itemTitle: 'Edit',
        handler:() => alert('edit clicked')
    },{
        itemTitle: 'Preview',
        handler:() => alert('preview clicked'),
    },{
        itemTitle: 'Expire',
        separator: true,
        handler:() => alert('expire clicked'),
    }];
```

## Events

|Attr|Description|
|---|---|
|`contextMenuOpened`|Triggered by opening the context menu|
|`contextMenuClosed`|Triggered by closing the context menu|
