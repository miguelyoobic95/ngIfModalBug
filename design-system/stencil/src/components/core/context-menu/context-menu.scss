
:host {
    display: block;
}

:host([hidden]) {
    display: none;
}

:host {
    .dropdown-toggle ::slotted(*) {
        cursor: pointer;
    }
    .dropdown-content {
        color: cssvar(dark-40);
        background-color: cssvar(light);
        width: 240px;
        max-width: 240px;
        box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
        visibility: hidden;
        /* hides sub-menu */
        opacity: 0;
        position: absolute;
        transform: translateY(-2em);
        z-index: -1;
        transition: all 0.3s ease-in-out 0s, visibility 0s linear 0.3s, z-index 0s linear 0.01s;
        max-height: 400px;
        border-radius: 0.25rem;
        .dropdown-entry {
            padding: 0.5rem 1rem;
            line-height: 1.5;
            width: 100%;
            cursor: pointer;
            &.border-top {
                border-top: 1px solid cssvar(dark-20);
            }
            &.border-bottom {
                border-top: 1px solid cssvar(dark-20);
            }
            &:hover {
                background: cssvar(accent-05);
                color: cssvar(dark);
            }
        }
    }
    .dropdown-content.show {
        visibility: visible;
        /* shows sub-menu */
        opacity: 1;
        z-index: 1;
        transform: translateY(0%);
        transition-delay: 0s, 0s, 0.3s;
        /* this removes the transition delay so the menu will be visible while the other styles transition */
    }
}