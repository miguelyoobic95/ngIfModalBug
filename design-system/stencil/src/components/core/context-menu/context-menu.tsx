import { Component, Prop, State, Method, Element, Event, EventEmitter } from '@stencil/core';
import { IContextMenuEntry } from '@shared/interfaces';

@Component({
    tag: 'yoo-context-menu',
    styleUrl: 'context-menu.scss',
    scoped: true
})
export class YooContextMenuComponent {

    @Prop() items: Array<IContextMenuEntry>;
    @Prop() context: any;

    @Event() contextMenuOpened: EventEmitter<boolean>;
    @Event() contextMenuClosed: EventEmitter<boolean>;

    @State() opened: boolean = false;

     @Element() host: HTMLStencilElement;

    @Method()
    open(): void {
        this.opened = true;
    }

    @Method()
    close(): void {
        this.opened = false;
    }

    componentWillLoad() {
        this.setupListener();
    }

    componentDidLoad() {
        this.calculateDropdownOpenDirection();
    }

    setupListener() {
        window.addEventListener('click', () => this.toggleWindow());
        window.addEventListener('touchstart', () => this.toggleWindow());
        window.addEventListener('resize', () => this.calculateDropdownOpenDirection());
    }

    calculateDropdownOpenDirection(): void {
        let rect = this.host.getBoundingClientRect();
        let position = window.innerHeight - rect.bottom;
        let dropdownContent = this.host.querySelector('.dropdown-content');
        let totalDropdownHeight = (rect.bottom - rect.top) + dropdownContent.clientHeight;
        dropdownContent.setAttribute('style', `transform: translateY(${dropdownContent.clientHeight > position ? '-' + totalDropdownHeight + 'px' : '0%'});`);
    }

    toggle(): void {
        if (!this.opened) {
            this.contextMenuOpened.emit(true);
            setTimeout(() => {
                this.open();
            }, 50);
        }
    }

    toggleWindow() {
        if (this.opened) {
            this.close();
            this.contextMenuClosed.emit(true);
        }
    }

    onItemClick(item: IContextMenuEntry, index: number): void {
        if (item && item.handler) {
            let context = this.context && this.context.toJS && item.sendImmutable !== true ? this.context.toJS() : this.context;
            item.handler(context, index);
        }
    }

    render(): JSX.Element {
        return [
            <span aria-haspopup="true" aria-expanded="false" onClick={() => this.toggle()}>
                <slot name="trigger"></slot>
            </span>,
            <div class={`${this.opened ? 'show' : ''} dropdown-content`}>
                {
                    this.items && this.items.length > 0 ?
                        this.items.map((item, i) =>
                            <div class={`${item.separator ? 'border-bottom' : ''} ${item.separatorAfter ? 'border-top' : ''} dropdown-entry`} onClick={() => this.onItemClick(item, i)}>{item.itemTitle}</div>)
                        : ''
                }
                <slot></slot>
            </div>
        ];
    }
}
