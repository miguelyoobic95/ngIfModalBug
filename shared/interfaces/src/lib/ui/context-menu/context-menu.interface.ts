export interface IContextMenuEntry {
    itemTitle?: string;
    icon?: string;
    class?: string;
    active?: boolean;
    separator?: boolean;
    separatorAfter?: boolean;
    hidden?: boolean;
    isHidden?: Function;
    visible?: boolean;
    handler?: Function;
    items?: Array<IContextMenuEntry>;
    sendImmutable?: boolean;
}
