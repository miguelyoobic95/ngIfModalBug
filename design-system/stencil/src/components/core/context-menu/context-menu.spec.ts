import { TestWindow } from '@stencil/core/testing';
import { YooContextMenuComponent } from './context-menu';
jest.useFakeTimers();

describe('YooContextMenuComponent', () => {
    const data = [{
        itemTitle: 'Edit',
        handler: () => alert('edit clicked')
    }, {
        itemTitle: 'Preview',
        handler: () => alert('preview clicked')
    }, {
        itemTitle: 'Expire',
        separator: true,
        handler: () => alert('expire clicked')
    }];
    it('should build', () => {
        expect(new YooContextMenuComponent()).toBeTruthy();
    });

    describe('Rendering', () => {
        it ('Should render a basic structure', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooContextMenuComponent],
                html: '<yoo-context-menu></yoo-context-menu>'
            });
            expect(element).toMatchSnapshot();
        });

        it('Should render with slots', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooContextMenuComponent],
                html : `<yoo-context-menu>
                            <yoo-button slot="trigger" text="test"></yoo-button>
                            <div>Lorem ipsum dolor sit amet consectetur adipisicing elit. Totam sequi vero laudantium cumque similique consectetur corporis blanditiis error neque ex libero, ipsa nostrum optio pariatur necessitatibus. Quas perferendis provident ipsam.</div>
                        </yoo-context-menu>`
            });
            expect(element).toMatchSnapshot();
        });

        it('Should render with items', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooContextMenuComponent],
                html: '<yoo-context-menu></yoo-context-menu>'
              });
            element.items = data;
            await window.flush();
            expect(element).toMatchSnapshot();
        });
    });

    describe('Methods', () => {
        it('onItemClick method should call item handler', async () => {
            let element = new YooContextMenuComponent();
            let item = {
                itemTitle: 'Edit',
                handler: jest.fn()
            };
            element.items = [item, {
                itemTitle: 'Preview',
                handler: () => alert('preview clicked')
            }, {
                itemTitle: 'Expire',
                separator: true,
                handler: () => alert('expire clicked')
            }];
            element.onItemClick(item, 0);
            expect(item.handler).toHaveBeenCalled();
        });
    });

    describe('events', () => {
        let element = new YooContextMenuComponent();
        element.contextMenuClosed = {
            emit: () => {}
        };
        element.contextMenuOpened = {
            emit: () => {}
        };

        it ('should emit open on button click and close on second button click', async () => {
            const openSpy = jest.spyOn(element.contextMenuOpened, 'emit');
            const closeSpy = jest.spyOn(element.contextMenuClosed, 'emit');
            element.items = data;
            element.toggle();
            expect(openSpy).toHaveBeenCalled();
            expect(setTimeout).toHaveBeenCalledTimes(1);
            jest.runOnlyPendingTimers();
            element.toggleWindow();
            expect(closeSpy).toHaveBeenCalled();
        });
    });
});
