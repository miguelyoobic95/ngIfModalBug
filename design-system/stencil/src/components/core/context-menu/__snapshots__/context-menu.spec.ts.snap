// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooContextMenuComponent Rendering Should render a basic structure 1`] = `
<yoo-context-menu
  class="hydrated"
  data-yoo-context-menu-host=""
>
  
  <span
    aria-expanded="false"
    aria-haspopup="true"
    data-yoo-context-menu=""
    data-yoo-context-menu-slot=""
  >
    
  </span>
  <div
    class="dropdown-content"
    data-yoo-context-menu=""
    data-yoo-context-menu-slot=""
    style="transform: translateY(0%);"
  >
    
    
  </div>
</yoo-context-menu>
`;

exports[`YooContextMenuComponent Rendering Should render with items 1`] = `
<yoo-context-menu
  class="hydrated"
  data-yoo-context-menu-host=""
>
  
  <span
    aria-expanded="false"
    aria-haspopup="true"
    data-yoo-context-menu=""
    data-yoo-context-menu-slot=""
  >
    
  </span>
  <div
    class="dropdown-content"
    data-yoo-context-menu=""
    data-yoo-context-menu-slot=""
    style="transform: translateY(0%);"
  >
    <div
      class="dropdown-entry"
      data-yoo-context-menu=""
    >
      Edit
    </div>
    <div
      class="dropdown-entry"
      data-yoo-context-menu=""
    >
      Preview
    </div>
    <div
      class="border-bottom dropdown-entry"
      data-yoo-context-menu=""
    >
      Expire
    </div>
    
  </div>
</yoo-context-menu>
`;

exports[`YooContextMenuComponent Rendering Should render with slots 1`] = `
<yoo-context-menu
  class="hydrated"
  data-yoo-context-menu-host=""
>
  
  
  
  
  
  
  <span
    aria-expanded="false"
    aria-haspopup="true"
    data-yoo-context-menu=""
    data-yoo-context-menu-slot=""
  >
    
    <yoo-button
      slot="trigger"
      text="test"
    />
  </span>
  <div
    class="dropdown-content"
    data-yoo-context-menu=""
    data-yoo-context-menu-slot=""
    style="transform: translateY(0%);"
  >
    
    
    
                            
    
                            
    <div>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Totam sequi vero laudantium cumque similique consectetur corporis blanditiis error neque ex libero, ipsa nostrum optio pariatur necessitatibus. Quas perferendis provident ipsam.
    </div>
    
                        
  </div>
</yoo-context-menu>
`;
